internal/promotion/runner/builtin/tar_extractor.go

```diff
@@ -0,0 +1,255 @@
+func (t *tarExtractor) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	// Validate the configuration against the JSON Schema.
+	if err := validate(t.schemaLoader, gojsonschema.NewGoLoader(stepCtx.Config), t.Name()); err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored}, err
+	}
+
+	// Convert the configuration into a typed object.
+	cfg, err := promotion.ConfigToStruct[builtin.UntarConfig](stepCtx.Config)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not convert config into %s config: %w", t.Name(), err)
+	}
+
+	return t.run(ctx, stepCtx, cfg)
+}
+
+func (t *tarExtractor) run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.UntarConfig,
+) (promotion.StepResult, error) {
+	// Secure join the paths to prevent path traversal attacks.
+	inPath, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.InPath)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not secure join inPath %q: %w", cfg.InPath, err)
+	}
+	outPath, err := securejoin.SecureJoin(stepCtx.WorkDir, cfg.OutPath)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not secure join outPath %q: %w", cfg.OutPath, err)
+	}
+
+	// Ensure the output directory exists
+	if err := os.MkdirAll(outPath, 0755); err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("failed to create output directory %q: %w", cfg.OutPath, err)
+	}
+
+	// Load the ignore rules.
+	matcher, err := t.loadIgnoreRules(outPath, cfg.Ignore)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("failed to load ignore rules: %w", err)
+	}
+
+	// Open the tar file
+	file, err := os.Open(inPath)
+	if err != nil {
+		return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("failed to open tar file %q: %w", cfg.InPath, err)
+	}
+	defer file.Close()
+
+	logger := logging.LoggerFromContext(ctx)
+	var tarReader *tar.Reader
+
+	// Check if the file is gzipped
+	gzr, err := gzip.NewReader(file)
+	if err == nil {
+		// File is gzipped
+		defer gzr.Close()
+		tarReader = tar.NewReader(gzr)
+		logger.Debug("treating file as gzipped tar")
+	} else {
+		// File is not gzipped, reset to beginning of file and try as regular tar
+		_, err = file.Seek(0, 0)
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("failed to seek in tar file: %w", err)
+		}
+		tarReader = tar.NewReader(file)
+		logger.Debug("treating file as regular tar")
+	}
+
+	// Extract the tar file
+	stripComponents := int64(0)
+	if cfg.StripComponents != nil {
+		stripComponents = *cfg.StripComponents
+		if stripComponents > 0 {
+			logger.Debug("stripping path components", "count", stripComponents)
+		}
+	}
+
+	for {
```

Comment:
We should ensure we cap the maximum size we allow to be extracted from the compressed artifact. This is something which was [recently fixed in Helm](https://github.com/helm/helm/commit/e4da49785aa6e6ee2b86efd5dd9e43400318262b), and is generally a [good practice to apply](https://github.com/fluxcd/pkg/blob/main/tar/tar.go#L28-L29) when multi-tenancy is handled within the same container.